var/spack/repos/builtin/packages/foam-extend/package.py

```diff
@@ -111,189 +107,77 @@ class FoamExtend(Package):
     depends_on('scotch~metis+mpi', when='+ptscotch')
     depends_on('metis@5:',         when='+metis')
     depends_on('parmetis',         when='+parmetis')
-    depends_on('parmgridgen',      when='+parmgridgen')
+    # mgridgen is statically linked
+    depends_on('parmgridgen',      when='+parmgridgen', type='build')
     depends_on('paraview@:5.0.1',  when='+paraview')
 
-    # Some user settings, to be adjusted manually or via variants
-    foam_cfg = {
-        'WM_COMPILER':        'Gcc',  # <- %compiler
-        'WM_ARCH_OPTION':     '64',   # (32/64-bit on x86_64)
-        # FUTURE? 'WM_LABEL_SIZE':      '32',  # <- +int64
-        'WM_PRECISION_OPTION': 'DP',  # <- +float32
-        'WM_COMPILE_OPTION':  'SPACKOpt',   # Do not change
-        'WM_MPLIB':           'USER',       # USER | USERMPI
-    }
+    # General patches
+    common = ['spack-Allwmake', 'README-spack']
+    assets = []
 
-    # The system description is frequently needed
-    foam_sys = {
-        'WM_ARCH':      None,
-        'WM_COMPILER':  None,
-        'WM_OPTIONS':   None,
+    # Some user config settings
+    config = {
+        'label-size': False,    # <- No int32/int64 support
+        'mplib': 'USERMPI',     # USER | USERMPI
     }
 
+    # The openfoam architecture, compiler information etc
+    _foam_arch = None
+
     # Content for etc/prefs.{csh,sh}
     etc_prefs = {}
 
     # Content for etc/config.{csh,sh}/ files
     etc_config = {}
 
-    build_script = './spack-Allwmake'  # <- Generated by patch() method.
-    # phases = ['configure', 'build', 'install']
-    # build_system_class = 'OpenfoamCom'
+    phases = ['configure', 'build', 'install']
+    build_script = './spack-Allwmake'  # <- Added by patch() method.
+
+    #
+    # - End of definitions / setup -
+    #
 
     def setup_environment(self, spack_env, run_env):
-        run_env.set('FOAM_INST_DIR', self.prefix)
+        run_env.set('FOAM_INST_DIR', os.path.dirname(self.projectdir)),
+        run_env.set('FOAM_PROJECT_DIR', self.projectdir)
         run_env.set('WM_PROJECT_DIR', self.projectdir)
-
-    @property
-    def _canonical(self):
-        """Canonical name for this package and version"""
-        return 'foam-extend-{0}'.format(self.version.up_to(2))
+        for d in ['wmake', self.archbin]:  # bin already added automatically
+            run_env.prepend_path('PATH', join_path(self.projectdir, d))
+        run_env.set('MPI_BUFFER_SIZE', "20000000")
+
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        """Provide location of the OpenFOAM project.
+        This is identical to the WM_PROJECT_DIR value, but we avoid that
+        variable since it would mask the normal OpenFOAM cleanup of
+        previous versions.
+        """
+        spack_env.set('FOAM_PROJECT_DIR', self.projectdir)
 
     @property
     def projectdir(self):
```

Comment:
Hi Adam,

For the top bit.
Yes I am an OpenFOAM developer (ESI-OpenCFD: www.openfoam.com).
Moving to another make system is fairly unlikely at this point. The hand-rolled `wmake`system is actually just a bunch of wrappers around a standard make, but obviously without all of the autotools or cmake configuration guessing. The main distinction though is probably that we have various lnInclude (symlinked include directory) with the headers for each major subsection. This means that the make file has the directory reference (via the `-I...` compiler option) without needing really long `#include ...` lines everywhere. If we go away from that, everything breaks and we have thousands of files...

OpenFOAM is indeed still a monster for using environment variables. The normal spack prefix is mostly okay, but I need to be able to build OpenFOAM with spack, but also be able to source the OpenFOAM environ and build depend modules without necessarily using spack.  For some (older) versions of OpenFOAM we some weird logic within OpenFOAM to find where the bashrc/cshrc files are located. I hope that we can make this even smoother in the future and drop this variable too.

Once the modules support fits better we should also be able to improve. If we can get this pull request merged in, any subsequent adjustments will presumably return the 'normal' range of 10-20 lines.

Cheers,
/mark